Modules/Scripted/Endoscopy/Endoscopy.py

```diff
@@ -834,11 +680,87 @@ def interpolateOrientations(self):
 
                 relativeOrientation = EndoscopyLogic.quaternionToOrientation(quaternion)
 
-                worldOrientation = self.relativeOrientationToWorld(self.resampledCurve, resampledCurvePointIndex, relativeOrientation)
+                worldOrientation = EndoscopyLogic.relativeOrientationToWorld(
+                    self.resampledCurve, resampledCurvePointIndex, relativeOrientation, self.planeNormal,
+                )
 
                 self.resampledCurve.SetNthControlPointOrientation(resampledCurvePointIndex, worldOrientation)
 
-    def getDefaultOrientation(self, curve, curveControlPointIndex):
+    def saveOrientationAtIndex(self, resampledCurvePointIndex, cameraNode):
+        # Compute CameraOrientations dictionary key
+        distanceAlongResampledCurve = EndoscopyLogic.distanceAlongCurveOfNthControlPoint(
+            self.resampledCurve, resampledCurvePointIndex,
+        )
+
+        # Compute CameraOrientations dictionary value
+        cameraPosition = cameraNode.GetPosition()
+        focalPoint = cameraNode.GetFocalPoint()
+        viewUp = cameraNode.GetViewUp()
+        worldMatrix3x3 = EndoscopyLogic.buildCameraMatrix3x3(cameraPosition, focalPoint, viewUp)
+        worldOrientation = EndoscopyLogic.matrix3x3ToOrientation(worldMatrix3x3)
+
+        # Add new dictionary key-value pair
+        cameraOrientations = EndoscopyLogic.getCameraOrientationsFromInputCurve(self.inputCurve)
+        cameraOrientations[distanceAlongResampledCurve] = worldOrientation
+
+        EndoscopyLogic.setInputCurveCameraOrientations(self.inputCurve, cameraOrientations)
+
+        return cameraOrientations
+
+    def removeOrientationAtIndex(self, resampledCurvePointIndexToDelete):
+        cameraOrientations = EndoscopyLogic.getCameraOrientationsFromInputCurve(self.inputCurve)
+
+        # We must freeze the cameraOrientations.keys() generator at the start (by converting it to a list) because we
+        # delete from the cameraOrientations Python dict in this loop.
+        for distanceAlongResampledCurve in list(cameraOrientations.keys()):
+            resampledCurvePointIndex = EndoscopyLogic.indexOfControlPointForDistanceAlongCurve(
+                self.resampledCurve, distanceAlongResampledCurve,
+            )
+            if resampledCurvePointIndex == resampledCurvePointIndexToDelete:
+                del cameraOrientations[distanceAlongResampledCurve]
+                break
```

Comment:
If the curve is edited, I suppose that it is possible that saved orientations for distinct keyframes now converge to the same keyframe.  (Or maybe that's not possible; I'm not sure.)  If it is possible, and the user asks to delete the orientation from a keyframe, do we want to leave ourselves in a state where there still is an orientation for that frame?  If we want to clear that frame completely then we should delete this `break` statement.

On the other hand, if we are looking for efficiency, we could terminate the loop once we're past the relevant keyframe; or instead do some sort of binary range search; ....

Reply:
Prior to this changes, we had the following:

```
        for distanceAlongResampledCurve in list(cameraOrientations.keys()):
            resampledCurvePointIndex = EndoscopyLogic.indexOfControlPointForDistanceAlongCurve(
                resampledCurve, distanceAlongResampledCurve,
            )
            if resampledCurvePointIndex == resampledCurvePointIndexToDelete:
                del cameraOrientations[distanceAlongResampledCurve]
```

My understanding is that for a given distance along the curve, we have one or zero saved orientation.

If we have a good reason to keep iterating once we found a matching orientation, the code should be documented to explain the rational.